scenario.py

```diff
@@ -189,33 +192,172 @@ def __init__(self, params, experiment_path):
+
+        # For each partner compose the list of clusters from which they will draw data samples
+        index = 0
+        for p in partners_with_specific_clusters:
+            p.clusters_list = labels[index:index + p.cluster_count]
+            index += p.cluster_count
+
+        shared_clusters = labels[index:index + shared_clusters_count]
+        for p in partners_with_shared_clusters:
+            p.clusters_list = random.sample(shared_clusters, k=p.cluster_count)
+
+        # We need to enforce the relative data amounts configured.
+        # It might not be possible to distribute all data samples, depending on...
+        # ... the coherence of the relative data amounts and the split option.
+        # We will compute a resize factor to determine the total nb of samples to be distributed per partner
+
+        # For partners getting data samples from specific clusters...
+        # ... compare the nb of available samples vs. the nb of samples initially configured
+        resize_factor_specific = 1
+        for p in partners_with_specific_clusters:
+            nb_available_samples = sum([count_per_cluster[cl] for cl in p.clusters_list])
+            nb_samples_configured = int(amounts_per_partner[p.id] * len(y_train))
+            ratio = nb_available_samples / nb_samples_configured
+            resize_factor_specific = min(resize_factor_specific, ratio)
+
+        # For each partner getting data samples from shared clusters:
+        # ... compute the nb of samples initially configured and resize it,
+        # ... then sum per cluster how many samples are needed.
+        # Then, find if a cluster is requested more samples than it got, and if yes by which factor
+        resize_factor_shared = 1
+        nb_samples_needed_per_cluster = dict.fromkeys(shared_clusters, 0)
+        for p in partners_with_shared_clusters:
+            initial_amount_resized = int(amounts_per_partner[p.id] * len(y_train) * resize_factor_specific)
+            initial_amount_resized_per_cluster = int(initial_amount_resized / p.cluster_count)
+            for cl in p.clusters_list:
+                nb_samples_needed_per_cluster[cl] += initial_amount_resized_per_cluster
+        for cl in nb_samples_needed_per_cluster:
+            resize_factor_shared = min(resize_factor_shared, count_per_cluster[cl] / nb_samples_needed_per_cluster[cl])
+
+        # Compute the final resize factor
+        final_resize_factor = resize_factor_specific * resize_factor_shared
+
+        # Size correctly each partner's subset. For each partner:
+        for p in partners_list:
+            p.final_nb_samples = int(amounts_per_partner[p.id] * len(y_train) * final_resize_factor)
+            p.final_nb_samples_p_cluster = int(p.final_nb_samples / p.cluster_count)
+        self.nb_samples_used = sum([p.final_nb_samples for p in partners_list])
+        final_relative_nb_samples = [round(p.final_nb_samples / self.nb_samples_used, 2) for p in partners_list]
```

Comment:
I would not use round here. We can round it later if we want to print it

Reply:
I would save this data in the results.csv file

Reply:
Ok